src/coreclr/jit/compiler.hpp

```diff
@@ -4703,6 +4703,10 @@ inline bool Compiler::compCanHavePatchpoints(const char** reason)
     {
         whyNot = "OSR can't handle reverse pinvoke";
     }
+    else if (!info.compIsStatic && !lvaIsOriginalThisReadOnly())
+    {
+        whyNot = "OSR can't handle modifiable this";
+    }
```

Comment:
What happens to these methods with tiering enabled? Do they end up in tier 0 or tier 1?
It seems risky if tail-recursive F# functions that are turned into loops by the compiler end up stuck in tier 0.

Reply:
`compCompileHelper` has the logic for this -- methods with explicit `tail` prefixes get optimized unless we're in PGO mode.

https://github.com/dotnet/runtime/blob/56a1cf4c892c6dec1d462e08e177f470e2f6b7b3/src/coreclr/jit/compiler.cpp#L6644-L6715

Reply:
They are not going to have any tail prefixes. I think #72845 is a case of a tail recursive function on lists that the F# compiler is turning into a loop in IL by reassigning `this`. These kind of methods may be quite common in F# code so getting them stuck in tier 0 could be problematic.

Reply:
If methods have explicit loops and modify `this` then they'll be disqualified from OSR by the logic being added to `compCanHavePatchpoints`, so will end up getting optimized.

For example, `Equals` in this test case now gets optimized immediately.

Reply:
Thanks, then it looks good.